pkg/kv/kvserver/rangefeed/rangefeed_datadriven_test.go

```diff
@@ -0,0 +1,1020 @@
+			nameToTs: make(map[string]hlc.Timestamp),
+			tsToName: make(map[hlc.Timestamp]string),
+		},
+		txns:     make(map[string]*kv.Txn),
+		startKey: testKey(sr),
+	}
+}
+
+func (e *env) close() {
+	for _, f := range e.feeds {
+		f.feed.Close()
+	}
+	e.tc.Stopper().Stop(context.Background())
+}
+
+func (e *env) setDefaultStepOptions() {
+	e.sortByTime = false
+	e.checkpointTimeout = 30 * time.Second
+}
+
+type timestamps struct {
+	nameToTs map[string]hlc.Timestamp
+	tsToName map[hlc.Timestamp]string
+}
+
+func (t timestamps) addTs(name string, ts hlc.Timestamp) (hlc.Timestamp, bool) {
+	// If name is used, return previous ts value to use instead.
+	if cts, ok := t.nameToTs[name]; ok {
+		return cts, cts.Equal(ts)
+	}
+	// If same timestamp is registered with different name, increment logical.
+	unchanged := true
+	if _, ok := t.tsToName[ts]; ok {
+		ts.Logical += 1
+		unchanged = false
+	}
+
+	t.nameToTs[name] = ts
+	t.tsToName[ts] = name
+	return ts, unchanged
+}
+
+func (t timestamps) addNextTs(ts hlc.Timestamp) {
+	name := fmt.Sprintf("ts%d", len(t.tsToName)+1)
+	t.addTs(name, ts)
+}
+
+func (t timestamps) getTsName(ts hlc.Timestamp) (string, bool) {
+	if n, ok := t.tsToName[ts]; ok {
+		return n, true
+	}
+	return "?", false
+}
+
+func (t timestamps) getTimestamp(name string) (hlc.Timestamp, bool) {
+	ts, ok := t.nameToTs[name]
+	return ts, ok
+}
+
+type testKey roachpb.Key
+
+func (p testKey) print(key roachpb.Key) string {
+	if len(key) > len(p) && key[:len(p)].Equal(roachpb.Key(p)) {
+		if key[len(key)-1] == 0 {
+			// This is generated end interval from some key, turn it into key+ to
+			// hint that this is past real key.
+			return string(key[len(p):len(key)-1]) + "+"
+		}
+		return string(key[len(p):])
+	}
+	return key.String()
+}
+
+func (p testKey) key(k string) roachpb.Key {
+	return roachpb.Key(append(p[:len(k):len(k)], k...))
+}
+
+type feedData struct {
+	t      *testing.T
+	span   roachpb.Span
+	dataMu struct {
+		syncutil.Mutex
+		values     eventStream
+		failure    error
+		failureSet bool
+	}
+	failedC chan interface{}
+
+	signalCP sync.Once
+	firstCP  chan interface{}
+
+	cpMu struct {
+		syncutil.Mutex
+		waiters []waitCP
+	}
+}
+
+func (d *feedData) onValue(_ context.Context, v *kvpb.RangeFeedValue) {
+	d.dataMu.Lock()
+	defer d.dataMu.Unlock()
+	d.dataMu.values = append(d.dataMu.values, testFeedEvent{v: v})
+	d.t.Logf("on Value: %s/'%s'", v.Key.String(), stringValue(v.Value.RawBytes))
+}
+
+func (d *feedData) onCheckpoint(_ context.Context, cp *kvpb.RangeFeedCheckpoint) {
+	d.dataMu.Lock()
+	d.dataMu.values = append(d.dataMu.values, testFeedEvent{cp: cp})
+	defer d.dataMu.Unlock()
+	d.signalCP.Do(func() {
+		close(d.firstCP)
+	})
+	d.t.Logf("on Checkpoint: [%s, %s) %s", cp.Span.Key.String(), cp.Span.EndKey.String(),
+		cp.ResolvedTS.String())
+	d.cpMu.Lock()
+	defer d.cpMu.Unlock()
+	for i, w := range d.cpMu.waiters {
+		if !w.targetTS.IsEmpty() {
+			if w.addCP(cp.ResolvedTS, cp.Span) {
+				close(w.timeReachedC)
```

Comment:
i'm a bit worried about the complexity of the rangefeed test driver code; in particular, what
happens if e.g. after sending checkpoint, rangefeed restarts (for whatever reasons -- and, btw -- it would be nice to add those restarts as a matter of test coverage) and after restart resends the same checkpoint... Then it's a race between this test completing and channel being closed twice.

I don't have a good answer here.  Perhaps taking rangefeed() method from dist_sender_rangefeed_test and foregoing the use of rangefeed library might make things simpler.  Perhaps not.  Is there anything that can be done to build up the confidence that this test isn't flaky even if some transient errors happen (and as I said above: it would be *great* to inject transient errors in the rangfeed event stream).